droidlet/dashboard/web/src/StateManager.js

```diff
@@ -380,6 +393,117 @@ class StateManager {
     interactionData[key] = value;
     this.socket.emit("interaction data", interactionData);
   }
+  
+  onObjectAnnotationSave(res) {
+    let { nameMap, pointMap, propertyMap } = res;
+    let newObjects = []
+    let scale = 500  // hardcoded from somewhere else
+    for (let id in nameMap) {
+      let oldObj = id < this.curFeedState.objects.length;
+      let newId = oldObj ? this.curFeedState.objects[id].id : null;
+      let newXyz = oldObj ? this.curFeedState.objects[id].xyz : null;
+      // Get rid of masks with <3 points
+      let i = 0
+      while (i < pointMap[id].length) {
+        if (!pointMap[id][i] || pointMap[id][i].length < 3) {
+          pointMap[id].splice(i, 1);
+          continue
+        }
+        i++
+      }
+      let newMask = pointMap[id].map(mask => mask.map((pt, i) => [pt.x * scale, pt.y * scale]))
+      let newBbox = this.getNewBbox(newMask);
+
+      newObjects.push({
+        label: nameMap[id], 
+        mask: newMask, 
+        properties: propertyMap[id].join("\n "),
+        type: "annotate", // either "annotate" or "detector"
+        id: newId, 
+        bbox: newBbox, 
+        xyz: newXyz, 
+      })
+    }
+    this.curFeedState.objects = newObjects
+
+    this.refs.forEach((ref) => {
+      if (ref instanceof LiveObjects) {
+        ref.setState({
+          objects: this.curFeedState.objects,
+        });
+      }
+    })
+  }
+
+  getNewBbox(maskSet) {
+    let xs = [], ys = []
+    for (let i = 0; i < maskSet.length; i++) {
+      for (let j = 0; j < maskSet[i].length; j++) {
+        xs.push(maskSet[i][j][0])
+        ys.push(maskSet[i][j][1])
+      }
+    }
+    let minX = Math.min.apply(null, xs), 
+        maxX = Math.max.apply(null, xs), 
+        minY = Math.min.apply(null, ys), 
+        maxY = Math.max.apply(null, ys)
+    return [minX, minY, maxX, maxY]
+  }
+
+  startLabelPropagation() {
+    let props = {
+      prevRgbImg: this.prevFeedState.rgbImg, 
+      depth: this.curFeedState.depth, 
+      prevDepth: this.prevFeedState.depth, 
+      prevObjects: this.prevFeedState.objects.filter(o => o.type === "annotate"), 
+      basePose: this.curFeedState.pose,
+      prevBasePose: this.prevFeedState.pose,
+    }
+    this.socket.emit("label_propagation", props)
+    // Reset
+    this.stateProcessed.rgbImg = true;
+    this.stateProcessed.depth = true;
+    this.stateProcessed.objects = true;
+    this.stateProcessed.pose = true;
+  }
+
+  labelPropagationReturn(res) {
+    this.refs.forEach((ref) => {
+      if (ref instanceof LiveObjects) {
+        for (let i = 0; i < res.length; i++) {
+          // Get rid of masks with <3 points
```

Comment:
I see this (getting rid masks with <3 points) done here and in `onObjectAnnotationSave` above. Why are both needed?

Reply:
One checks points generated by the label prop script and the other checks points generated by the annotator tool. Although we already have a check in the annotator tool, so I'll get rid of one of these checks.